diff --git hmc/puppet/modules/hdp-ganglia/manifests/monitor.pp hmc/puppet/modules/hdp-ganglia/manifests/monitor.pp
index e20533d..48224db 100644
--- hmc/puppet/modules/hdp-ganglia/manifests/monitor.pp
+++ hmc/puppet/modules/hdp-ganglia/manifests/monitor.pp
@@ -8,15 +8,17 @@ class hdp-ganglia::monitor(
   if ($service_state == 'no_op') {
   } elsif ($service_state in ['running','stopped','installed_and_configured','uninstalled']) {
     if ($monitor_and_server_single_node == false) {
-      include hdp-ganglia #note: includes the common package ganglia-monitor
+      #note: includes the common package ganglia-monitor
+      include hdp-ganglia 
       class { 'hdp-ganglia::config': 
         ganglia_server_host => $ganglia_server_host,
         require             => Class['hdp-ganglia'],
         before              => Class['hdp-ganglia::monitor::config-gen']
       }
     }
-
-    class { 'hdp-ganglia::monitor::config-gen': }
+    anchor {'hdp-ganglia::monitor::begin' : } ->
+    class { 'hdp-ganglia::monitor::config-gen': } ->
+    anchor {'hdp-ganglia::monitor::end' : } 
 
     if ($monitor_and_server_single_node == false) {
       Class['hdp-ganglia'] -> Class['hdp-ganglia::monitor::config-gen']
@@ -41,7 +43,7 @@ class hdp-ganglia::monitor::config-gen()
   if ($hdp-ganglia::params::omit_namenode != true) {
     hdp-ganglia::config::generate_monitor { 'HDPNameNode':}
   }
-  if ($hdp-ganglia::params::omit_job_tracker != true) {
+  if ($hdp-ganglia::params::omit_jobtracker != true) {
     hdp-ganglia::config::generate_monitor { 'HDPJobTracker':}
   }
   if ($hdp-ganglia::params::omit_hbase_master != true) {
diff --git hmc/puppet/modules/hdp-ganglia/manifests/params.pp hmc/puppet/modules/hdp-ganglia/manifests/params.pp
index aeb3e40..a60e947 100644
--- hmc/puppet/modules/hdp-ganglia/manifests/params.pp
+++ hmc/puppet/modules/hdp-ganglia/manifests/params.pp
@@ -13,7 +13,7 @@ class hdp-ganglia::params() inherits hdp::params
 
 
   $omit_hbase_master = hdp_default("omit_hbase_master",hdp_is_empty($hdp::params::hbase_master_host))
-  $omit_job_tracker = hdp_default("omit_job_tracker",hdp_is_empty($hdp::params::jtnode_host))
+  $omit_jobtracker = hdp_default("omit_jobtracker",hdp_is_empty($hdp::params::jtnode_host))
   $omit_namenode = hdp_default("omit_namenode",hdp_is_empty($hdp::params::namenode_host))
   $omit_slaves = hdp_default("omit_slaves",hdp_is_empty($hdp::params::slave_hosts))
 }
\ No newline at end of file
diff --git hmc/puppet/modules/hdp-ganglia/manifests/server.pp hmc/puppet/modules/hdp-ganglia/manifests/server.pp
index 91d1e6d..bfafc18 100644
--- hmc/puppet/modules/hdp-ganglia/manifests/server.pp
+++ hmc/puppet/modules/hdp-ganglia/manifests/server.pp
@@ -24,7 +24,7 @@ class hdp-ganglia::server(
         ganglia_service => 'gmond'
       }
     }
-    if ($hdp-ganglia::params::omit_job_tracker != true) {
+    if ($hdp-ganglia::params::omit_jobtracker != true) {
       hdp-ganglia::config::generate_server { 'HDPJobTracker':
         ganglia_service => 'gmond'
       }
@@ -56,9 +56,9 @@ class hdp-ganglia::server(
       Class['hdp-ganglia'] -> Class['hdp-ganglia::server::packages']
     }
 
-    Class['hdp-ganglia::server::packages'] -> Class['hdp-ganglia::config'] -> 
-      Hdp-ganglia::Config::Generate_server<||> -> Class['hdp-ganglia::server::services']
-  } else {
+    anchor{'hdp-ganglia::server::begin':} -> Class['hdp-ganglia::server::packages'] -> Class['hdp-ganglia::config'] -> 
+      Hdp-ganglia::Config::Generate_server<||> -> Class['hdp-ganglia::server::services'] -> anchor{'hdp-ganglia::server::end':}
+   } else {
     hdp_fail("TODO not implemented yet: service_state = ${service_state}")
   }
 }
